src/vdpcmp.1
.TH "vdpcmp" 1
.SH NAME
vdpcmp \- Compare two images using Visual Difference Predicator (low level command)
.SH SYNOPSIS
.B vdpcmp
<target_file> <mask_file> [--ldr]  [--mask <val>] [--psycho <val>] [--detection-mechanism <dmID>] [--no-phase-uncertainty] [--dump <pattern>] [--verbose] [--help] 
.SH DESCRIPTION
This command processes a pair of images using Visual Difference
Predicator and writes a map of probability detection to standard
output. It is more convenient to use 'vdp' instead of this command, as
it can handle image format conversion.

\fB<target_file>\fR is the file name of an image for which detection
probability should be computed (e.g. distorted image) and
\fB<mask_file>\fR is the file name of an image that is a reference
(e.g. original, not distorted). The order in which both images are
passed to the vdp is important and may affect the results.

The format of input images and output map is pfs. Use one of the
readers from pfstools package to read images into vdpcmp.

.SH OPTIONS

.TP
.B \--ldr

Use the original (non high dynamic range) version of the VDP. Since
the original VDP has not been maintained for ages, this option is
\fIdepreciated\fR. HDR VDP should provide better results and can work with
LDR images as well.

.TP
.B \--multiply-lum <val>

Multiply luminance values of both target and mask by <val>. Using such
multiplier is necessary if the images contain only relative but not
absolute luminance values (cd/m^2). In case of low dynamic range
images that have been converted from sRGB color space to XYZ
(\fB--linear\fR switch of \fBpfsinppm\fR), \fB--multiply-lum\fR should
be set to the maximum luminance of the display (see also man page of
\fBvdp\fR).

.TP
.B \--mask <val>
Set the slope of the masking function to <val>. Default 1.


.TP
.B \--psycho <val>
Set the slope of the psychometric function to <val>. Default 3.5.

.TP
.B \--peak-contrast <contrast>
The peak sensitivity of the visual system given as minimum
distinguishable contrast. The value must be given as a fraction, not
percent!. Default value: 0.006 (=0.6%)

.TP
.B \--otf <otfID>
Name of the optical tranfer function (OTF, or PSF of the eye). Possible
values:

.B DEELEY 
(default) - OTF from 'Deeley, R.J., Drasdo, N., & Charman,
W. N. (1991). A simple parametric model of the human ocular modulation
transfer function. Ophthalmology and Physiological Optics, 11,
91-93'. It takes into account pupil size, which depends on adaptation
luminance. This is suggested OTF function for the VDP.

.B MARIMONT
- OTF from 'Marimont D.H. and Wandell B.A., 1994. Matching colour
images; the effects of axial chromatic aberration. Journal of the
Optical Society of America A, 11 (no 12) 2113-3122'. It takes into
account pupil size, which depends on adaptation luminance. (note: the
integration performed in the current implementation is not
accurate. This can lead to artifacts in the filtered image).

.B SPENCER
- PSF from Spencer et al. SIGGRAPH'95 paper 

.B NORMANN
- PSF from Normann et al. IEEE Trans. on Systems, Man and Cybernetics '83 paper (note: this function can be wrong (typo in the paper?), therefore it is not recommended to use it).

.B WESTHEIMER
- PSF from Westheimer "The eye as an optical instrument" in Handbook of Perception and Visual Performance. This function is also used in Sarnoff Visual Discrimination Model.

.B NONE
- no OTF preprecessing 


.TP
.B \--detection-mechanism <dmID>
Name of the detection mechanism to use. Currently supported:

.B WATSON, VDP
- Watson cortex transform (default)

.B SIMPLE
- no multichannel processing, only a single channel

.TP
.B \--no-abs-fix

VDP can predict visible differences for pixels that are identical
because of inaccuracies of visual band filters. To avoid this, as a
final step HDR VDP compares two input images pixel by pixel and if the
corresponding pixels are almost identical (difference < 0.0001), it
sets the prediction for that pixel to 0. Use \fB--no-abs-fix\fR to
disable this behavior.

.TP
.B \--dump <pattern>
Dump intermediate images resulting from VDP processing. The pattern
specifies names of the files to dump. A shell pattern ('*' is a wild
card character) can be used to dump many frames. The option can be
specified many times to dump additional images. All images are dumped
in pfs format. Some available dump images are listed below. More can
be found in the sources.
.RS
.PD 0
.TP
.B target_otf.pfs
target after OTF filtering
.TP
.B mask_otf.pfs
mask after OTF filtering
.TP
.B filter_otf.pfs
OTF filter
.TP
.B filter_csf_N.pfs
CSF filter for the adaptation luminance 10^N
.TP
.B adaptation_map.pfs
adaptatin map used for CSF filtering
.TP
.B target_photoreceptor_nonlin.pfs
target after photoreceptor nonlinearity
.TP
.B mask_photoreceptor_nonlin.pfs
mask after photoreceptor nonlinearity
.TP
.B target_csf.pfs
target after CSF filtering
.TP
.B mask_csf.pfs
mask after CSF filtering
.TP
.B dom_filter_cortex_N.pfs
DOM filter for the band N (1-6)
.TP
.B mask_cortex_N_M.pfs
mask for the band N and orientatiom M after cortex transform
.TP
.B target_cortex_N_M.pfs
target for the band N and orientatiom M after cortex transform
.TP
.B filter_cortex_N_M.pfs
cortex filter for the band N and orientatiom M
.TP
.B difference_cortex_N_M.pfs
difference between target and mask for the band N and orientatiom M
.TP
.B te_cortex_N_M.pfs
threshold elevation (due to masking) for the band N and orientatiom M
.TP
.B probability_cortex_N_M.pfs
probability of detection for the band N and orientatiom M
.RE
.PD
.TP
.B \--no-phase-uncertainty
Phase uncertainty improves the prediction of visual masking by
discounting the influence of phase from the masking thresholds. By
default, phase uncertainty is included in the VDP calculation. However
it almost doubles time required to process an image. This switch can
be used to turn off computation of phase uncertainty.

.PP
The following options define viewing conditions. Because passing a
complete set of options to specify viewing conditions is rather
tedious, the better way is to use \fB-vc\fR option of the \fIvdp\fR
command (see the man page for vdp).

.TP
.B \--display-width \--display-height
Physical dimensions of the display given in meters. Default: width
0.375 height 0.300.

.TP
.B \--min-distance \--max-distance
Minimum and maximum distance from which the display is
observed. Default: min 0.5 max 0.5.

.TP
.B \--display-x-resolution \--display-y-resolution
Vertical and horizontal resolution of the display given in
pixels. Default: horizontal 1280 vertical 1024.

.TP
.B \--pixels-per-degree ppd
The number of pixels per 1 visual degree. Use this option instead of display-width/height and display-x/y-resolution.

.SH "SEE ALSO"
.BR vdp (1)
.BR vdpvis (1)
.BR pfsin (1)
.SH BUGS
Please report bugs and comments to Rafal Mantiuk
<mantiuk@mpi-sb.mpg.de>.


ChangeLog
        * Fixed:  Configure script handles -enable-debug with overwriting
	CXXFLAGS 

	* Added: New option: pixels-per-degree
	
v1.7.1 2010-12-19  Rafal Mantiuk  <mantiuk@gmail.com>

        * Added: Documented color coding in the manual of vdpvis
	
	* Added: bash auto-detection in configure.ac
	
	* Fixed: pupil diameter formula (previous fix was wrong)
	
	* Fixed: vdp script recognizes BMP files and is run with bash instead of sh

	* Fixed: pupil diameter replaced with pupil radius in Merimont &
	Wandel's OTF (thanks to Kenichiro Masaoka)

v1.7 2008-08-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Fixed: spurious zeros in the fan filters (thanks to Kenichiro Masaoka)
	
	* Fixed: frequency misalignment bug (thanks to Kenichiro Masaoka)

	* Fixed: pupil diameter formula 
	
	* Added: support for multithreaded FFTW (though I could not observe any speed up)	
	
	* Added: --sumary option to vdpvis (dump percent summary to a text
	file)

	* Fixed: target and mask image order was wrong in some programs

	* Fixed: if OTF is used, NTF (neural) is used instead of CSF
		
v1.6 2006-06-13  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Fixed: previous version could return different results on
	differents runs because fftw library used different IFFT code,
	although the difference was minimal. To make VDP results
	deterministic, FFTW optimizations are switched to FFTW_ESTIMATE.	

v1.5 2006-02-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Added: ABS fix, which compares target and mask and zeros
	predictions if the pixels are almost identical

	
v1.4 2005-12-02  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Added: Remove Inf, Nan and nonpositive values from luminance
	channels before any processing

v1.3 2005-05-09  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Low dynamic range images are handled automatically by vdp script

	* Fixed to work with pfstools 1.2
	
	* Added switch --multiply-lum
	
	* Added more comments and copyright information

v1.2 2004-11-22  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Added new OTFs: Marimont, Deeley (suggested)
	
	* Added option to the vdp script: -o <raw_vdp_output.pfs>
	
	* The default parameters set to the results of calibration
	procedure from SPIE'05 paper

v1.1 2004-09-23  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	 
	* Added: phase uncertainty. Now the VDP works twice as slower as
	it requires additional filtering step for each band. However this
	should predict masking more acuratelly.
	
	* Fixed: Original VDP parameters set to Daly's settings
	
	* Fixed: Peak sensitivity of HDR VDP changed to 0.005% contrast,
	to match the original VDP
	
	* Added: possibility to specify viewing conditions (options or
	config file)
	 
v1.0 2004-09-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Initial release

	  


src/vdpcmp.cpp
/**
 * @brief main procedure of the HDR VDP, command line interface
 * 
 * This file is a part of HDR VDP.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 *
 * @author Rafal Mantiuk, <mantiuk@mpi-inf.mpg.de>
 *
 * $Id: vdpcmp.cpp,v 1.4 2010/12/25 16:56:06 rafm Exp $
 */

#include <iostream>
#include <stdlib.h>
#include <stdio.h>
#include "array2d_algorithm.h"

#include <pfs.h>
#include <getopt.h>
#include <math.h>

#include "otf.h"
#include "nonlinearity.h"
#include "fftw_array2d.h"
#include "csf_filter.h"
#include "cortex_transform.h"
#include "fftutils.h"

#include "dump_image_aspect.h"

#define PROG_NAME "vdpcmp"

class QuietException 
{
};

#include <assert.h>

void fixPredictionWithRMS( const pfs::Array2D *inTarget, const pfs::Array2D *inMask, pfs::Array2D *probabilityMap );

void newProcessVDP( const pfs::Array2D *target, const pfs::Array2D *mask,
  pfs::Array2D *probabilityMap,
  OTF *otf,
  PhotoreceptorNonlin *nonlinearity, CSFFilter *csfFilter, DetectionMechanism *dm,
  bool fixRMS );

#if 0
// This piece of code is used for debugging purposes only.

#define PFSEOL "\x0a"
static void dumpPFS( const char *fileName, const pfs::Array2D *data, const char *channelName )
{
  FILE *fh = fopen( fileName, "wb" );
  assert( fh != NULL );

  int width = data->getCols();
  int height = data->getRows();

  fprintf( fh, "PFS1" PFSEOL "%d %d" PFSEOL "1" PFSEOL "0" PFSEOL
    "%s" PFSEOL "0" PFSEOL "ENDH", width, height, channelName );

  for( int y = 0; y < height; y++ )
    for( int x = 0; x < width; x++ ) {
      fwrite( &((*data)(x,y)), sizeof( float ), 1, fh );
    }
  
  fclose( fh );
}
static void dumpPFS( const char *fileName, FFTWComplexArray *data, const char *channelName )
{
  FILE *fh = fopen( fileName, "wb" );
  assert( fh != NULL );

  int width = data->getFFTCols();
  int height = data->getFFTRows();

  fprintf( fh, "PFS1" PFSEOL "%d %d" PFSEOL "1" PFSEOL "0" PFSEOL
    "%s" PFSEOL "0" PFSEOL "ENDH", width, height, channelName );

  const fftwf_complex *rawData = data->getData();
  
  for( int y = 0; y < height; y++ )
    for( int x = 0; x < width; x++ ) {
      float mod = sqrt( rawData[x+y*width][0]*rawData[x+y*width][0] +
        rawData[x+y*width][1]*rawData[x+y*width][1] );
      fwrite( &mod, sizeof( float ), 1, fh );
    }
  
  fclose( fh );
}

unsigned long checkSum( const pfs::Array2D *data )
{
  const int pixelCount = data->getRows()*data->getCols();
  unsigned long sum = 0;
  for( int i = 0; i < pixelCount; i++ ) {
    float v = (*data)(i);
    sum = sum ^ *((unsigned long*)&v);    
  }
  return sum;
}

unsigned long checkSum( const FFTWComplexArray *data )
{
  const int pixelCount = data->getRows()*data->getCols();
  unsigned long sum = 0;
  const fftwf_complex *d = data->getData();
  
  for( int i = 0; i < pixelCount; i++ ) {
    float v = d[i][0];
    sum = sum ^ *((unsigned long*)&v);    
  }
  return sum;
}
#endif

static void errorCheck( bool condition, const char *string )
{
  if( !condition ) {
    throw pfs::Exception( string );
  }
}

void printHelp()
{
  fprintf( stderr, PROG_NAME " <mask_file> <target_file> [--ldr] [--mask <val>] [--psycho <val>] [--otf <otfID>] [--peak-contrast <contrast>] [--detection-mechanism <dmID>] [--no-phase-uncertainty] [--dump <pattern>] [--display-x-resolution] [--display-y-resolution] [--display-width] [--display-height] [--min-distance] [--max-distance] [--pixels-per-degree <pixels>] [--multiply-lum] [--no-abs-fix] [--verbose] [--help]\n"
    "See man page for more information.\n" );
}

void processVDP(int argc, char **argv)
{
  const char *targetFile = NULL, *maskFile = NULL;
  float maskingSlope = 1;
  float psychometricFunctionSlope = 3.5; // The value from the patent
  float multiplyLum = 1;
  const char *detectionMechID = NULL;
  const char *nonlinearityID = NULL;
  const char *otfID = "DEELEY";
  bool hdr = true;
  bool phaseUncertainty = true;
  float minDistance = 0.5, maxDistance = 0.5;
  float displayWidth = 0.375, displayHeight = 0.300;
  int displayXResolution = 1280, displayYResolution = 1024;
  float peakContrast = 0.006;
  float pixelsPerDegree = -1;
  bool absFix = true;

  bool verbose = false;

  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "mask", required_argument, NULL, 'm' },
    { "psycho", required_argument, NULL, 'p' },
    { "detection-mechanism", required_argument, NULL, 't' },
    { "otf", required_argument, NULL, 'o' },    
    { "nonlinearity", required_argument, NULL, 'n' }, // TODO
    { "ldr", no_argument, NULL, 'l' }, 
    { "no-phase-uncertainty", no_argument, NULL, 'u' }, 
    { "no-pu", no_argument, NULL, 'u' }, 
    { "dump", required_argument, NULL, 'd' },
    { "display-x-resolution", required_argument, NULL, 'x' },
    { "display-y-resolution", required_argument, NULL, 'y' },
    { "display-width", required_argument, NULL, 'w' },
    { "display-height", required_argument, NULL, 'e' },
    { "min-distance", required_argument, NULL, 'i' },
    { "max-distance", required_argument, NULL, 'a' },
    { "peak-contrast", required_argument, NULL, 'c' },
    { "multiply-lum", required_argument, NULL, '*' },
    { "no-abs-fix", no_argument, NULL, 'R' },
    { "pixels-per-degree", required_argument, NULL, 'r' },
    { NULL, 0, NULL, 0 }
  };

  initializeDumpImageAspect();
  
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, "m:p:vr:", cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'm':
      maskingSlope = (float)strtod( optarg, NULL );
      break;
    case 'p':
      psychometricFunctionSlope = (float)strtod( optarg, NULL );
      break;
    case 'd':
      dumpImageAspect->addIncludePattern( optarg );
      break;
    case 't':
      detectionMechID = optarg;
      break;
    case 'o':
      otfID = optarg;
      break;
    case 'u':
      phaseUncertainty = false;
      break;
    case 'l':
      hdr = false;
      break;
    case 'x':
      displayXResolution = strtol( optarg, NULL, 10 );
      break;
    case 'y':
      displayYResolution = strtol( optarg, NULL, 10 );
      break;
    case 'w':
      displayWidth = (float)strtod( optarg, NULL );
      break;
    case 'e':
      displayHeight = (float)strtod( optarg, NULL );
      break;
    case 'i':
      minDistance = (float)strtod( optarg, NULL );
      break;      
    case 'a':
      maxDistance = (float)strtod( optarg, NULL );
      break;      
    case 'c':
      peakContrast = (float)strtod( optarg, NULL );
      break;
    case 'r':
      pixelsPerDegree = (float)strtod( optarg, NULL );
      break;
    case 'R':
      absFix = false;
      break;
    case '*':
      multiplyLum = (float)strtod( optarg, NULL );
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  } 

  
//    int oi = optind;
//    while( oi < argc )
//      cerr << "option: " << argv[oi++] << "\n";
  
  // Exactly two non-option arguments: target, and mask
  errorCheck( optind >= (argc - 2), "Too many input arguments" );
  errorCheck( optind == (argc - 2), "Both target and mask files must be specified" );

  targetFile = argv[optind++];
  maskFile = argv[optind++];

  if( verbose ) {
    std::cerr << "Psychometric function slope: " << psychometricFunctionSlope << "\n";
    std::cerr << "Threshold elevation slope: " << maskingSlope << "\n";
  }
  
  FILE *targetFh;
  FILE *maskFh;

  bool targetIsStdin = !strcmp( "-", targetFile );
  bool maskIsStdin = !strcmp( "-", maskFile );

  errorCheck( !(targetIsStdin && maskIsStdin),
    "Only one of the target and mask pair of images can be loaded from standard input" );  
  
  targetFh = targetIsStdin ? stdin : fopen( targetFile, "rb" );
  errorCheck( targetFh != NULL, "Can not open target image file" );
  
  maskFh = maskIsStdin ? stdin : fopen( maskFile, "rb" );
  errorCheck( maskFh != NULL, "Can not open mask image file" );

  errorCheck( pixelsPerDegree == -1 || pixelsPerDegree > 0, "pixels-per-degree value must be positive" );
  
  pfs::DOMIO pfsio;
    
  // Load target image (with artifacts)
  pfs::Frame *targetFrame = pfsio.readFrame( targetFh );
  pfs::Array2D *targetY = targetFrame->getChannel( "Y" );
  errorCheck( targetY != NULL, "Luminance channel missing in the target image" );
  clampToPositive( targetY, targetY ); // Remove <=0 and nan(s)
    
  // Load mask image (original)
  pfs::Frame *maskFrame = pfsio.readFrame( maskFh );
  pfs::Array2D *maskY = maskFrame->getChannel( "Y" );
  errorCheck( maskY != NULL, "Luminance channel missing in the mask image" );
  clampToPositive( maskY, maskY ); // Remove <=0 and nan(s)
  
  // Check if sizes match
  errorCheck( maskY->getRows() == targetY->getRows() &&
    maskY->getCols() == targetY->getCols(), "Can compare only images of the same size" );

  // Multiply luminance if necessary
  if( multiplyLum != 1 ) {
    const int pixels = maskY->getRows()*maskY->getCols();
    for( int i = 0; i < pixels; i++ ) {
      (*maskY)(i) *= multiplyLum;
      (*targetY)(i) *= multiplyLum;      
    }
  }
  
  // VDP processing blocks
  OTF *otf = NULL;
  PhotoreceptorNonlin *nonlinearity = NULL;
  CSFFilter *csfFilter = NULL;
  DetectionMechanism *dm = NULL;
  
  {                           // Configure processing blocks of the VDP
    float contrastNormalizationFactor = 0; // Used if nonlinearity == photoreceptor
    ViewingConditions viewCond( displayXResolution, displayYResolution,
      displayWidth, displayHeight, minDistance, maxDistance );

    if( pixelsPerDegree != -1 ) {
      viewCond = ViewingConditions( pixelsPerDegree, minDistance, maxDistance );
    } 
      

    if( verbose )
      fprintf( stderr, "Pixels per degree: %g\n", viewCond.getPixelsPerDegree( (viewCond.minDistance + viewCond.maxDistance)/2 ) );
    
    
    if( hdr == false ) {        // Use original Daly's VDP
      nonlinearity = new VDPNonlinearity( );
      //Normalization factor used to compute contrast between mask and
      //target Sugested values are display device mean and logAverage(
      //maskY ), but those do not give stable results. This
      //implementation follows the patent and set normalization factor
      //to 1 (no normalization)
      contrastNormalizationFactor = 1; // Should be 
      fprintf( stderr, "conrast norm: %g\n", contrastNormalizationFactor );
      csfFilter = new VDPCSF( maskY->getCols(), maskY->getRows(), viewCond, 30 );
      otfID="NONE";
    }    

    if( !strcasecmp( otfID, "SPENCER" ) ) {
      otf = new SpencerOTF( viewCond, maskY->getCols(), maskY->getRows(),
        logAverage( maskY ) );
    } else if( !strcasecmp( otfID, "NORMANN" ) ) {
      otf = new NormannBaxterOTF( viewCond, maskY->getCols(), maskY->getRows() );
    } else if( !strcasecmp( otfID, "WESTHEIMER" ) ) {
      otf = new WestheimerOTF( viewCond, maskY->getCols(), maskY->getRows() );
    } else if( !strcasecmp( otfID, "MARIMONT" ) ) {
      otf = new MarimontOTF( viewCond, maskY->getCols(), maskY->getRows(), logAverage( maskY ) );
    } else if( !strcasecmp( otfID, "DEELEY" ) ) {
      otf = new DeeleyOTF( viewCond, maskY->getCols(), maskY->getRows(), logAverage( maskY ) );
    } else if( !strcasecmp( otfID, "NONE" ) ) {
      otf = NULL;
    } else {
      throw pfs::Exception( "Unknown optical transfer function (OTF). Possible values: MARIMONT, DEELEY, SPENCER, NORMANN, WESTHEIMER, NONE (default)" );
    }    
    
    if( nonlinearity == NULL )
      nonlinearity = new JNDScaledNonlinearity( peakContrast );
    if( csfFilter == NULL ) {
      csfFilter = new MultiAdaptationCSF( maskY->getCols(), maskY->getRows(), viewCond,
        (otf == NULL ? NULL : otf->getFilter() ) );
    }
    if( detectionMechID == NULL || !strcasecmp( detectionMechID, "VDP" ) ||
      !strcasecmp( detectionMechID, "WATSON" ) ) { // default dm
      dm = new WatsonDM( psychometricFunctionSlope, maskingSlope, contrastNormalizationFactor != 0, contrastNormalizationFactor, phaseUncertainty );
    } else if( !strcasecmp( detectionMechID, "SIMPLE" ) ) {
      dm = new SingleChannelDM( psychometricFunctionSlope );
    } else {
      throw pfs::Exception( "Unknown detection mechanism. Possible values: VDP, WATSON, SIMPLE" );
    }
  }
  
  // Prepare output for saving
  const int cols = targetY->getCols(), rows = targetY->getRows();
  pfs::Channel *resultMap = targetFrame->createChannel( "VDP" );
  
  newProcessVDP( targetY, maskY, resultMap, otf, nonlinearity, csfFilter, dm, absFix );

  // Write output
  pfsio.writeFrame( targetFrame, stdout );
  pfsio.freeFrame( targetFrame );
  pfsio.freeFrame( maskFrame );
  
  delete nonlinearity;
  delete csfFilter;
  delete dm;
  
  if( targetFh != stdin ) fclose( targetFh );
  if( maskFh != stdin ) fclose( maskFh );  
  
  
}

int main( int argc, char* argv[] )
{
  try {
    processVDP( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}

void newProcessVDP( const pfs::Array2D *inTarget, const pfs::Array2D *inMask,
  pfs::Array2D *probabilityMap,
  OTF *otf,
  PhotoreceptorNonlin *nonlinearity, CSFFilter *csfFilter, DetectionMechanism *dm,
  bool absFix )
{

#ifdef HAVE_LIBFFTW_THREADS
  // This should make things faster on 2 and more CPU machines, but I
  // did not observe any performance improvement
  fftwf_init_threads();
  fftwf_plan_with_nthreads(2);
#endif
  
  const int rows = inTarget->getRows(), cols = inTarget->getCols();
  
  BidomainArray2D target( cols, rows ), mask( cols, rows );

  // Copy images to structures that can handle fft
  pfs::copyArray( inTarget, target.setSpatial() );
  pfs::copyArray( inMask, mask.setSpatial() );
  
  if( otf != NULL ) {
    std::cerr << "OTF: " << otf->getName() << "...\n";
    otf->process( &target, &target );
    otf->process( &mask, &mask );
  }
  
  dumpImageAspect->dump( "target_otf.pfs", &target, "Y" );
  dumpImageAspect->dump( "mask_otf.pfs", &mask, "Y" );

  BidomainArray2D adaptationMap( cols, rows );
  adaptationMap = mask;
  dumpImageAspect->dump( "adaptation_map.pfs", &adaptationMap, "Y" );
  
  std::cerr << "Nonlinearity: " << nonlinearity->getName() << "...\n";
  
  nonlinearity->process( &target, &target );
  nonlinearity->process( &mask, &mask );

  dumpImageAspect->dump( "target_photoreceptor_nonlin.pfs", &target, "Y" );
  dumpImageAspect->dump( "mask_photoreceptor_nonlin.pfs", &mask, "Y" );

//   FFTWComplexArray freqTarget( cols, rows );
//   FFTWComplexArray freqMask( cols, rows );

  std::cerr << "CSF filtering: " << csfFilter->getName() << "...\n";
  
  csfFilter->process( &target, &target, &adaptationMap );
  csfFilter->process( &mask, &mask, &adaptationMap );

  dumpImageAspect->dump( "target_csf.pfs", &target, "Y" );
  dumpImageAspect->dump( "mask_csf.pfs", &mask, "Y" );
  
  dm->process( &target, &mask, probabilityMap );

  // Compare target and mask and zero predictions if pixels are
  // almost identical
  if( absFix ) {
    std::cerr << "Applying ABS fix\n";
    fixPredictionWithRMS( inTarget, inMask, probabilityMap );
  }

#ifdef HAVE_LIBFFTW_THREADS
  fftwf_cleanup_threads();
#endif
  
}

/**
 * Compare target and mask and zero predictions if pixels are almost
 * identical
 */
void fixPredictionWithRMS( const pfs::Array2D *inTarget, const pfs::Array2D *inMask, pfs::Array2D *probabilityMap ) 
{
  const int pixels = inTarget->getCols()*inTarget->getRows();

  for( int i = 0; i < pixels; i++ ) {
    float d = fabsf( (*inTarget)(i) - (*inMask)(i) );
    if( d < 0.0001 ) (*probabilityMap)(i) = 0;
  }
  
}
